openstack_plugin/resources/compute/server.py

```diff
@@ -510,28 +516,88 @@ def _get_port_networks(client_config, port_ids):
     :param dict client_config: Openstack configuration required to connect
     to API
     :param (List) port_ids: List of uuid ports
-    :return:
+    :return: Dict networks: contains map between port_id & network_id
     """
-    networks = []
-    for port_id in port_ids:
+    def _get_network(port_id):
         port = OpenstackPort(client_config=client_config, logger=ctx.logger)
         port.resource_id = port_id
         response = port.get()
-        networks.append(response.network_id)
+        return {
+            'uuid': response.network_id,
+            'port': port_id
+        }
+    return map(_get_network, port_ids)
 
-    return networks
 
+def _remove_duplicated_nics_from_relationships(nics_from_rels, client_config):
 
-def _clean_duplicate_networks(server_config):
-    """
-    This method will clean all duplicates network items from server config
-    :param dict server_config: The server configuration required in order to
-    create the server instance using Openstack API
+    # Get the ports from relationships if they are existed
+    port_ids = find_openstack_ids_of_connected_nodes_by_openstack_type(
+        ctx, PORT_OPENSTACK_TYPE)
+
+    # Need to check if nics relationships contains ports that connect to the
+    # networks which are part of the networks associated with the current
+    # relationship, so basically it is not allowed to have ports connected
+    # to networks already exist in the server relationships
+
+    # In order to solve this issue, the following actions are required:
+    # 1. Get the associated network for port and update "nics_from_rels" list
+    # 2. Clean the "nics_from_rels" to remove any duplicates entries that
+    # have the same network object (maintains orders)
+    port_networks = _get_port_networks(client_config, port_ids)
+    port_nic = {}
+    inverted_port_nic = {}
+    # Convert port related to networks
+    for port_network in port_networks:
+        for network in nics_from_rels:
+            if network.get('uuid') == port_network.get('uuid'):
+                # Replace the port id with network id so that we can remove
+                # the duplicates and maintain the orders
+                port_nic[port_network['uuid']] = port_network['port']
+    if port_nic:
+        inverted_port_nic = dict(map(reversed, port_nic.items()))
+
+    unique_set = set()
+    ordered_list = []
+    if port_nic:
+        for item in nics_from_rels:
+            data = tuple(sorted(item.items()))
+            if data[0][1] in port_nic.keys():
+                data = data + (('port', port_nic[data[0][1]]),)
+            elif data[0][1] in port_nic.values():
+                data = (('uuid', inverted_port_nic[data[0][1]]),) + data
+            if data not in unique_set:
+                unique_set.add(data)
+                ordered_list.append(dict(data))
```

Comment:
Can we change the name of "data" to something more readable? :)

Reply:
Yup